prov/gni/include/gnix_vector.h

```diff
@@ -0,0 +1,456 @@
+ * @var attr	The attributes of this vector.
+ * @var ops	The supported operations on this vector.
+ * @var vector	The begging address of the vector.
+ * @var size	The current size of the vector.
+ * @var lock	A read/write lock for the vector.
+ */
+typedef struct gnix_vector {
+	gnix_vec_state_e state;
+	gnix_vec_attr_t attr;
+	gnix_vector_ops_t *ops;
+	gnix_vec_entry_t *vector;
+	rwlock_t lock;
+} gnix_vector_t;
+
+
+/*******************************************************************************
+ * API Prototypes
+ ******************************************************************************/
+/**
+ * Create the initial vector.  The user is responsible for initializing the
+ * "attr" parameter prior to calling this function.
+ *
+ * @param[in] vec	the vector to initialize
+ * @param[in] attr	the vector attributes
+ *
+ * @return FI_SUCCESS	Upon successfully creating the vector
+ * @return -FI_EINVAL	Upon receiving an invalid parameter
+ * @return -FI_ENOMEM	Upon insufficient memory to create the vector
+ */
+int _gnix_vec_init(struct gnix_vector *vec, gnix_vec_attr_t *attr);
+
+/**
+ * Close the vector elements and then the vector.
+ *
+ * @param[in] vec       the vector to close
+ *
+ * @return FI_SUCCESS	Upon successfully closing the vector
+ * @return -FI_EINVAL	Upon a uninitialized or dead vector
+ */
+int _gnix_vec_close(gnix_vector_t *vec);
+
+/**
+ * Resize the vector to size.
+ *
+ * @param[in] vec	the vector to resize
+ * @param[in] size	the new size of the vector
+ *
+ * @return FI_SUCCESS	Upon successfully resizing the vector
+ * @return -FI_EINVAL	Upon passing a uninitialized or dead vector, a size
+ * 			less than the minimum vector size, or a size greater
+ *			than the maximum vector size
+ * @return -FI_ENOMEM	Upon running out of memory
+ */
+int _gnix_vec_resize(gnix_vector_t *vec, uint64_t size);
+
+/*******************************************************************************
+ * internal helper fns
+ ******************************************************************************/
+static inline uint64_t __gnix_vec_get_new_size(gnix_vector_t *vec, uint64_t index)
+{
+	uint64_t new_size = vec->attr.cur_size;
+
+	if (vec->attr.vec_increase_type == GNIX_VEC_INCREASE_ADD) {
+		do {
+			new_size += vec->attr.vec_increase_step;
+		} while (index >= new_size);
+	} else {
+		if (new_size)
+			new_size *= vec->attr.vec_increase_step;
+		else
+			new_size = (new_size + 1) * vec->attr.vec_increase_step;
+
+		while (index >= new_size)
+			new_size *= vec->attr.vec_increase_step;
+
+	}
+
+	if (new_size > vec->attr.vec_maximum_size) {
+		GNIX_WARN(FI_LOG_EP_CTRL, "Maximum vector size of %lu "
+			  "reached in __gnix_vec_new_size\n", vec->attr.vec_maximum_size);
+		new_size = vec->attr.vec_maximum_size;
+	}
+
+	return new_size;
+}
+
+static inline int __gnix_vec_resize(gnix_vector_t *vec, uint64_t new_size)
+{
+	vec->vector = realloc(vec->vector, new_size * sizeof(gnix_vec_entry_t));
+
+	if (!vec->vector) {
```

Comment:
If there is not enough space, it returns null, so this check is somewhat appropriate. However, if you want to maintain the existing data, I recommend storing the returned pointer in a temporary variable. By storing the function result in vec->vector, failed realloc calls will also cause the original pointer to be lost. This might be picked up on valgrind in very rare cases.